README.md
# rcrimeanalysis: An Implementation of Crime Analysis Methods <a><img src='../master/images/rcrimeanalysis-hex.png' align="right" height="139" /></a>

<!-- badges: start -->
![CRAN Version](https://www.r-pkg.org/badges/version-ago/rcrimeanalysis)
![CRAN Downloads](https://cranlogs.r-pkg.org/badges/last-month/rcrimeanalysis)
![Build Status](https://travis-ci.org/JSSpaulding/rcrimeanalysis.svg?branch=master)
<!-- badges: end -->

## Overview

**rcrimeanalysis** is a package containing various functions for the analysis of crime incident or records management system (RMS) data. The package implements analysis algorithms scaled for city or regional crime analysis units including kernel density estimation for crime heat maps, geocoding using the 'Google Maps' API, identification of repeat crime incidents, spatio-temporal map comparison across time intervals, time series analysis (forecasting and decomposition), detection of optimal parameters for the identification of near repeat incidents, and near repeat analysis with crime network linkage. The package also contains an example dataset of Chicago crime incident data. The incidents were acquired from the [Chicago Data Portal](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2). From these data, the set was subset to 25000 incidents of various primary types from 2017 to 2019.

**Keywords**: Crime Analysis, Geographic Information Systems (GIS), Spatio-Temporal, Near Repeat Analysis, Open Source

## Getting Started

These instructions will allow any user to install the R package on your local machine for development and testing purposes. 

### Installing

Begin by installing the package from either CRAN or the GitHub repo:

```
install.packages('rcrimeanalysis') #CRAN version
devtools::install_github('JSSpaulding/rcrimeanalysis') #Dev version
```

After installation, load and attach the package

```
library(rcrimeanalysis)
```

The following will call the example Chicago Data Portal crime incident data from the package. Inspection of the data is crucial because the package functions are designed for use of this data structure.

```
data(crimes)
```

## Usage

The following examples illustrate the usage of the functions within the **rcrimeanalysis** package.

### geocode_address

The *geocode_address* function leverages the utilities of the **ggmap** package for the batch geocoding of physical addresses. Geocoding is an essential element of crime analysis where locations are converted into grid coordinates for mapping and other analysis. Note that this function uses the Google Maps API which requires Google Cloud Credentials (available at [https://cloud.google.com/maps-platform/](https://cloud.google.com/maps-platform/). Also see the [ggmap package](https://github.com/dkahle/ggmap) for details.

```
library(ggmap)
register_google("**Google Cloud Credentials Here**")
addresses <- c("Milan Puskar Stadium, Morgantown, WV","Woodburn Hall, Morgantown, WV")
test1 <- geocode_address(addresses)

```

### id_repeat
This function identifies crime incidents which occur at the same location and returns a list of such incidents where each data frame in the list contains the RMS data for the repeat crime incidents. The data is based on the Chicago Police Department RMS structure.

```
repeat_out <- id_repeat(crimes)
repeat_out[[1]] #prints first repeat crime incident table
```

### kde_int_comp

The *kde_int_comp* function calculates and compares the kernel density estimate (heat maps) of crime incident locations from two given intervals. The function returns a net difference raster illustrating changes between the spatial crime distributions across the specified intervals for comparion. This is also useful in identifying crime displacement. Additionally, a shiny.tag.list object which contains three leaflet widgets: a widget with the calculated KDE from interval 1, a widget with the calculated KDE from interval 2, and a widget with a raster of the net differences between the KDE (heat maps) of each specified interval is returned for interactive visualization of the results. The net difference object utilizes an implementation of the parula color pallete for optimal visualization.

```
int_out <- kde_int_comp(crimes, start1="1/1/2017", end1="3/1/2017",
                                start2="1/1/2018", end2="3/1/2018")
int_out
```

An example of the shiny.tag.list object is given below for the above example.

<a><img src='../master/images/kde-interval1.PNG' align="center" height="400" /></a>

### kde_map

The *kde_map* function computes a kernel density estimate of crime incident locations and returns a **Leaflet** map of the incidents. The input data structure is based on the Chicago Police Department RMS (example data) and populates pop-up windows with the incident location for each incident. The resultant map has three layers: an ESRI base-map, all crime incidents plotted (with incident info pop-up windows), and a kernel density estimate of those points. The user can also specify the 'pts' parameter which controls whether the crime incidents are plotted.

```
library(leaflet) #needed to install basemap providers
library(leafsync) #creates plot of both maps
crime_sample <- head(crimes, n = 1000)
# Plot without Points
p1 <- crime_sample %>% kde_map(pts = FALSE)
# Plot with Incident Points
p2 <- crime_sample %>% kde_map()
leafsync::sync(p1,p2)
```

An example of the resultant map is given below. A zoomed in portion is also shown (right) to illustrate the incident data pop-up. 

<a><img src='../master/images/kde-map1.png' align="center" height="400" /></a>

### near_repeat_analysis

The **near_repeat_analysis** function performs near repeat analysis on a set of incident locations. The user specifies distance and time thresholds which are utilized to search all other incidents and find other incidents within these parameters. The function returns a list of all identified near repeat series within the input data as **igraph** graph objects. The output list can be used to render linkage plots of each series and to discern the near repeat linkages between the crime incidents. The function extends traditional near repeat analysis to also provide linkages which identify potential series of incidents. Note that this function is dependent on the memory of the system, it may be necessary to subset datasets prior to use. 

```
nr_data <- head(crimes, n = 5000) #truncate dataset for near repeat analysis
out <- near_repeat_analysis(data = nr_data, tz = "America/Chicago", epsg = "32616")
path <- paste0(getwd(), "/netout") #path for iGraph networks out
name <- 1
# Save Image of Each igraph Network to Netpath Directory
library(igraph)
for(i in out){
    png(file = paste(path, "/series", name, ".png", sep = ""))
    plot(i, layout = layout_with_lgl, edge.color="orange",
    vertex.color = "orange", vertex.frame.color = "#ffffff",
    vertex.label.color = "black")
    dev.off()
    name <- name + 1
}
```

### near_repeat_eval
This function performs an evaluation of given crime incidents to reccomend parameters for near repeat analysis. A series of time and distance parameters are tested using a full factorial design using the set of incident locations to determine the frequency of occurrence given each set of parameters. The results of the full factorial assessment are then modeled through interpolation and the second derivative is calculated to determine the inflection point. The inflection point represents the change in frequency of detected incidents which near repeat. Determination of the inflection point is completed for both the time and distance domains.

```
data(crimes)
nr_dat <- subset(crimes, crimes$primary_type == "BURGLARY")
pars <- near_repeat_eval(data=nr_dat, tz="America/Chicago", epsg="32616")
pars
```

### ts_daily_decomp

The **ts_daily_decomp** function transforms daily crime count data and plots the resultant components of a time series which has been decomposed into seasonal, trend, and irregular components using Loess smoothing. Holt Winters exponential smoothing is also performed for inproved trend resolution since data is in a daily format. This is valuable to understand changes in crime frequency over time for policy evaluation, tactical effectiveness, and detection of potential influx of organized crime activities.

```
test <- ts_daily_decomp(crimes, start = c(2017, 1, 1))
plot(test)
```

### ts_forecast

The **ts_forecast** function transforms traditional crime data into a time series and forecasts future incident counts based on the input data over a specified duration. The forecast is computed using simple exponential smoothing with additive errors. Returned is a plot of the time series, trend, and the upper and lower prediction limits for the forecast. Such a prediciton is useful for administrative planning and allocation of resources. 

```
library(prophet)
ts_forecast(crimes, start = c(2017, 1, 1))
```

An example forecast plot is given below for the above example.

<a><img src='../master/images/ts-forecast.png' align="center" height="500" /></a>


### ts_monthly_decomp

The **ts_monthly_decomp** function transforms traditional crime data and plots the resultant components of a time series which has been decomposed into seasonal, trend and irregular components using Loess smoothing.

```
test <- ts_month_decomp(crimes, 2017)
plot(test)
```

An example decomposition plot is given below for homicides from the *crimes* dataset.

<a><img src='../master/images/homicide-ts.png' align="center" height="500" /></a>

## Built With

* [ggmap](https://github.com/dkahle/ggmap) - Integration of Google Maps API
* [igraph](https://igraph.org/r/) - Package used for plotting near repeat network series
* [leaflet](https://github.com/Leaflet/Leaflet) - JavaScript library for mobile-friendly interactive maps


## Authors

* **Jamie Spaulding** - *Author, Contributor*

* **Keith Morris** - *Contributor*


NEWS.md
# rcrimeanalysis (Development Version)

* Development version has no changes from 0.4.3.

# rcrimeanalysis 0.5.0

* Removed package dependency on `rgdal` functionality. `rgdal` set to be archived Oct. 2023. `near_repeat_analysis()` and `near_repeat_eval()` were changed to depend on 'terra' for gdal integration. 

* Vignettes updated.

# rcrimeanalysis 0.4.2

* Added /vignettes with examples of package functionality.

* Added NEWS.md for entire package history.

* Added hex sticker for package.

# rcrimeanalysis 0.4.1

* Added `near_repeat_eval()` function to accompany `near_repeat_analysis()`. The function determines the ideal parameters to use for near repeat analysis given a full factorial assessment of spatio-temporal clustering in the provided dataset.

# rcrimeanalysis 0.4.0

* Update of `ts_daily_decomp()`, `ts_monthly_decomp()`, and `ts_forecast()` to remove `Prophet` dependency. These functions now utilize Seasonal Decomposition Of Time Series By Loess Smoothing functionality to decompose and forecast of the input time series

# rcrimeanalysis 0.3.1

* Example correction for `geocode_address()` with minor bug fixes.

# rcrimeanalysis 0.3.0

* `id_repeat()` function added to detect repeat crime incidents based on location. Outputs a list of repeat crimes by location.

* Change `kde_int_comp()` output from static raster to interactive leafsync widget with three maps for overall better results visualization. 

* Added the 'pts' parameter to `kde_map()` which specifies whether crime incident points are to be included in the rendered KDE crime map. Default parameter setting is `pts = TRUE`.

# rcrimeanalysis 0.2.0

* Update `geocode_address()` with API fix.

* Update README.md and overall package descriptions.

# rcrimeanalysis 0.1.0

* Initial CRAN Version. 

FUNCTIONS IN INITAL VERSION

* `geocode_address()`

* `kde_int_comp()`

* `kde_map()`

* `near_repeat_analysis()`

* `ts_daily_decomp()`

* `ts_monthly_decomp()`

* `ts_forecast()`


R/near_repeat_eval.R
## near_repeat_eval
## Jamie Spaulding

#' Identification of Optimal Time and Distance Parameters for Near Repeat Analysis
#' @description This function performs an evaluation of given crime incidents to
#'     reccomend parameters for near repeat analysis. A series of time and distance
#'     parameters are tested using a full factorial design using the set of
#'     incident locations to determine the frequency of occurrence given each
#'     set of parameters. The results of the full factorial assessment are then
#'     modeled through interpolation and the second derivative is calculated to
#'     determine the inflection point. The inflection point represents the
#'     change in frequency of detected incidents which near repeat. Determination
#'     of the inflection point is completed for both the time and distance domains.
#' @param data Data frame of crime or RMS data. See provided Chicago Data Portal
#'     example for reference
#' @param epsg The EPSG Geodetic Parameter code for the area being considered.
#'     The EPSG code is used for identifying projections and performing coordinate
#'     transformations. If needed, the EPSG for an area can be found at
#'     \url{https://spatialreference.org}.
#' @param tz Time zone for which the area being examined. By default this value
#'     is assigned as the same time zone of the system. For more information
#'     about time zones within R, see \url{https://www.rdocumentation.org/packages/base/versions/3.6.1/topics/timezones}.
#' @return Returns a data frame with one instance (row) of two fields (columns).
#'     The fields are: distance and time. The instance indicates the optimal
#'     near repeat parameters for each. Note that distance is given in meters
#'     and time is given as days.
#' @author Jamie Spaulding, Keith Morris
#' @keywords spatial
#' @examples
#' \dontshow{
#' data(crimes)
#' nr_dat <- head(subset(crimes, crimes$primary_type == "BURGLARY"), n = 100)
#' pars <- near_repeat_eval(data=nr_dat, tz="America/Chicago", epsg="32616")
#' }
#' \donttest{
#' data(crimes)
#' nr_dat <- subset(crimes, crimes$primary_type == "BURGLARY")
#' pars <- near_repeat_eval(data=nr_dat, tz="America/Chicago", epsg="32616")
#' pars
#' }
#' @importFrom igraph graph_from_adjacency_matrix
#' @importFrom igraph components
#' @importFrom stats approx
#' @importFrom stats complete.cases
#' @importFrom stats dist
#' @importFrom terra crds
#' @importFrom terra crs
#' @importFrom terra vect
#' @importFrom utils txtProgressBar
#' @importFrom utils setTxtProgressBar
#' @export
near_repeat_eval <- function(data, epsg, tz=NULL){
  # Set Run Sequence of Parameters to Evaluate -----
  day_interval <- c(0, 0.25, 0.5, 1, 2, 3, 4, 5, 6, 7, 10, 14, 21, 28, 180, 365)
  dist_interval <- c(0, 1, 5, 10, 50, 100, 250, 500, 750, 1000, 2000, 5000)
  run_seq <- expand.grid(day_interval,dist_interval)
  names(run_seq) <- c("TimeThresh", "DistThresh")

  # Set Defaults -----
  if (is.null(tz)) {tz <- Sys.timezone(location = TRUE)} #default: system location
  crs <- paste0("+init=epsg:", as.character(epsg))

  # Date Formats -----
  data$datetime <- as.POSIXct(data$date, tz = tz, "%m/%d/%Y %H:%M") #date-time object
  data$date <- as.Date(data$date, "%m/%d/%Y %H:%M") #ensure date column is in Date format
  crime <- data[stats::complete.cases(data), ] #only complete cases

  # Parameter Evaluation -----
  series_num <- NULL
  jj <- 1
  pb = utils::txtProgressBar(min = 0, max = nrow(run_seq), initial = 0, style = 3)
  for(i in 1:nrow(run_seq)) {
    utils::setTxtProgressBar(pb,i)
    a <- run_seq[i,]
    DistThresh <- a[,1]
    TimeThresh <- a[,2]

    # Set Coordinate Reference System ----
    vcoord <- terra::vect(cbind(crime$longitude, crime$latitude))
    terra::crs(vcoord) <- crs
    coordsout <- terra::crds(vcoord)
    crime$x1 <- coordsout[,1] #bind coordinate 1 to crime data
    crime$x2 <- coordsout[,2] #bind coordinate 2 to crime data

    # Near Repeat Analysis using Threshold Parameters -----
    SpatDist <- as.matrix(stats::dist(crime[,c('x1','x2')])) < DistThresh
    TimeDist <- as.matrix(stats::dist(crime$date)) < TimeThresh
    AdjMat <- SpatDist * TimeDist #under both distance and under time
    row.names(AdjMat) <- crime$case_number #case numbers for labels in igraph
    colnames(AdjMat) <- crime$case_number #case numbers for labels in igraph

    # igraph network from adjacency matrix -----
    G <- igraph::graph_from_adjacency_matrix(AdjMat, mode="undirected", diag = FALSE)
    CompInfo <- igraph::components(G) #assigning the connected components
    out <- data.frame(CompId=CompInfo$membership, CompNum=CompInfo$csize[CompInfo$membership])
    out <- out[out$CompNum!=1, ] #remove any series consisting of 1 incident
    series_num[jj] <- nrow(out)
    jj <- jj+1
  }

  eval_out <- cbind(run_seq,series_num)

  # Interpolate Evaluation Results -----
  x1_interp <- stats::approx(eval_out$TimeThresh, eval_out$series_num, ties = mean)
  datx1 <- data.frame(x1 = x1_interp[[1]], y = x1_interp[[2]])

  x2_interp <- stats::approx(eval_out$DistThresh, eval_out$series_num, ties = mean)
  datx2 <- data.frame(x2 = x2_interp[[1]], y = x2_interp[[2]])

  # Calculate Time First Derivative -----
  dy <- NULL
  dx <- NULL
  for (i in 2:nrow(datx1)) {
    dy[i] <- datx1$y[i] - datx1$y[i - 1]
    dx[i] <- datx1$x1[i] - datx1$x1[i - 1]
  }
  first <- dy/dx
  datx1.1 <- data.frame(x = datx1$x1, y = first)

  # Calculate Time Second Derivative -----
  datx1.1[1,2] <- 0
  dy2 <- NULL
  dx2 <- NULL
  for (i in 2:nrow(datx1.1)) {
    dy2[i] <- datx1.1$y[i] - datx1.1$y[i - 1]
    dx2[i] <- datx1.1$x[i] - datx1.1$x[i - 1]
  }
  second <- dy2/dx2

  data <- data.frame(x = datx1.1$x, y = second)
  time_out <- floor(data[which.min(data$y),1]) #Optimal Distance Parameter

  # Calculate Distance First Derivative -----
  dy <- NULL
  dx <- NULL
  for (i in 2:nrow(datx2)) {
    dy[i] <- datx2$y[i] - datx2$y[i - 1]
    dx[i] <- datx2$x2[i] - datx2$x2[i - 1]
  }
  first <- dy/dx
  datx2.1 <- data.frame(x = datx2$x2, y = first)

  # Calculate Distance Second Derivative -----
  datx2.1[1,2] <- 0
  dy2 <- NULL
  dx2 <- NULL
  for (i in 2:nrow(datx2.1)) {
    dy2[i] <- datx2.1$y[i] - datx2.1$y[i - 1]
    dx2[i] <- datx2.1$x[i] - datx2.1$x[i - 1]
  }
  second <- dy2/dx2
  data2 <- data.frame(x = datx2.1$x, y = second)
  dist_out <- floor(data2[which.min(data2$y),1]) #Optimal Dist Parameter

  results <- data.frame(distance = dist_out, time = time_out)
  return(results)
}
